File: src/hp_guard/matching.py
from __future__ import annotations
from .models import Rule, PolicyCall


def rule_matches_rule(rule: Rule, call: PolicyCall) -> bool:
    for field_name, expected_value in rule.target.items():
        if field_name in {"agent", "tool", "user"}:
            if expected_value != getattr(call, field_name):
                return False
        elif field_name.startswith("context."):
            context_key = field_name.removeprefix("context.")
            if call.context.get(context_key) != expected_value:
                return False
        else:
            return False
    return True

File: src/matching.rs
use crate::models::{PolicyCall, Rule};

/// Check whether a rule's target fields are compatible with a PolicyCall.
/// All specified target fields are ANDed. Unspecified fields match anything.
pub fn rule_matches_rule(rule: &Rule, call: &PolicyCall) -> bool {
    for (key, value) in &rule.target {
        match key.as_str() {
            "agent" => match &call.agent {
                Some(call_agent) if call_agent != value => return false,
                None => return false,
                _ => {}
            },
            "tool" => match &call.tool {
                Some(call_tool) if call_tool != value => return false,
                None => return false,
                _ => {}
            },
            "user" => match &call.user {
                Some(call_user) if call_user != value => return false,
                None => return false,
                _ => {}
            },
            _ => {
                // Unknown target keys cause the rule to NOT match
                return false;
            }
        }
    }
    true
}
